pkg/lib/dialogs.jsx

```diff
@@ -0,0 +1,72 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2022 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* DIALOG PRESENTATION PROTOCOL
+ *
+ * Example:
+ *
+ * import { WithDialogs, useDialogs } from "dialogs.jsx";
+ *
+ * const App = () =>
+ *   <WithDialogs>
+ *     <Page>
+ *       <ExampleButton />
+ *     </Page>
+ *   </WithDialogs>;
+ *
+ * const ExampleButton = () => {
+ *   const D = useDialogs();
+ *   return <Button onClick={() => D.show(<MyDialog />)}>Open dialog</Button>;
+ * };
+ *
+ * const MyDialog = () => {
+ *   const D = useDialogs();
+ *   return (
+ *     <Modal title="My dialog"
+ *            isOpen
+ *            onClose={D.close}>
+ *       <p>Hello!</p>
+ *     </Modal>);
+ * };
+ *
+ * This does two things: It maintains the state of whether the dialog
+ * is open, and it does it high up in the DOM, in a stable place.
+ * Even if ExampleButton is no longer part of the DOM, the dialog will
+ * stay open and remain useable.
+ */
+
+import React, { useState, useContext } from "react";
+
+export const DialogsContext = React.createContext();
+export const useDialogs = () => useContext(DialogsContext);
+
+export const WithDialogs = ({ children }) => {
+    const [dialog, setDialog] = useState(null);
+
+    const D = {
```

Comment:
Could this be called `DialogContext` or `RCDialog` ("react context") or so? In this library the name doesn't matter much, but having almost-global `D` variables in our page code is hard to read and prone to collisions.

Reply:
Yes, I was thinking too that just `D` is too cute... :-)  It's just a pattern, every component can decide by itself what it wants to call this thing, but we should establish a proper pattern from the start.

What about `Dialogs`?  `const Dialogs = useDialogs();`